overlord/snapshotstate/snapshotstate.go

```diff
@@ -0,0 +1,271 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snapshotstate
+
+import (
+	"fmt"
+	"sort"
+
+	"golang.org/x/net/context"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/overlord/snapshotstate/backend"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/strutil"
+)
+
+var (
+	snapstateAll                     = snapstate.All
+	snapstateCheckChangeConflictMany = snapstate.CheckChangeConflictMany
+	backendIter                      = backend.Iter
+)
+
+type snapshotSetState struct {
+	LastID uint64 `json:"last-id"`
+}
+
+func newSnapshotSetID(st *state.State) (uint64, error) {
+	var setState snapshotSetState
+
+	err := st.Get("snapshot-set", &setState)
+	if err == state.ErrNoState {
+		setState = snapshotSetState{}
+	} else if err != nil {
+		return 0, err
+	}
+
+	setState.LastID++
+	st.Set("snapshot-set", setState)
+
+	return setState.LastID, nil
+}
+
+func allActiveSnapNames(st *state.State) ([]string, error) {
+	all, err := snapstateAll(st)
+	if err != nil {
+		return nil, err
+	}
+	names := make([]string, 0, len(all))
+	for name, snapst := range all {
+		if snapst.IsInstalled() {
+			names = append(names, name)
+		}
+	}
+
+	sort.Strings(names)
+
+	return names, nil
+}
+
+func snapNamesInSnapshotSet(setID uint64, requested []string) (snapsFound []string, filenames []string, err error) {
+	sort.Strings(requested)
+	found := false
+	err = backendIter(context.Background(), func(r *backend.Reader) error {
+		if r.SetID == setID {
+			found = true
+			if len(requested) == 0 || strutil.SortedListContains(requested, r.Snap) {
+				snapsFound = append(snapsFound, r.Snap)
+				filenames = append(filenames, r.Name())
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	if !found {
+		return nil, nil, client.ErrSnapshotSetNotFound
+	}
+	if len(snapsFound) == 0 {
+		return nil, nil, client.ErrSnapshotSnapsNotFound
+	}
+
+	return snapsFound, filenames, nil
+}
+
+// checkSnapshotChangeConflict checks whether there's an in-progress change for snapshots with the given set id.
+func checkSnapshotChangeConflict(st *state.State, setID uint64, conflictingKinds ...string) error {
+	for _, task := range st.Tasks() {
```

Comment:
bit unclear why we don't go over changes then tasks here, given that the first set of checks are about full changes, also the comment of this should probably say this is also change kind based, is not that common